File: model/src/main/java/io/github/runtimemodels/chazm/entity/AgentEntity.java
package io.github.runtimemodels.chazm.entity;

import com.google.inject.assistedinject.Assisted;
import io.github.runtimemodels.chazm.id.UniqueId;
import lombok.Getter;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

class AgentEntity extends AbstractEntity<Agent> implements Agent {

    @Getter
    private final ContactInfo contactInfo;

    @Inject
    AgentEntity(@NotNull @Assisted final UniqueId<Agent> id, @NotNull @Assisted final ContactInfo contactInfo) {
        super(id);
        this.contactInfo = contactInfo;
    }

    @Override
    public boolean equals(final Object object) {
        if (object instanceof Agent) {
            final Agent agent = (Agent) object;
            return super.equals(agent);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public String toString() {
        return super.toString();
    }

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/AgentEntity.kt
package runtimemodels.chazm.model.entity

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.Agent
import runtimemodels.chazm.api.id.AgentId
import javax.inject.Inject

internal data class AgentEntity @Inject constructor(
    @Assisted override val id: AgentId
) : Agent {
    internal data class DefaultContactInfo(
        private val map: MutableMap<Any, Any> = mutableMapOf()
    ) : Agent.ContactInfo, Map<Any, Any> by map {
        override fun add(key: Any, value: Any) {
            map[key] = value
        }

        override fun remove(key: Any) {
            map.remove(key)
        }
    }

    override val contactInfo: Agent.ContactInfo = DefaultContactInfo()
}
